optuna/visualization/_pareto_front.py

```diff
@@ -171,53 +130,200 @@ def _make_scatter_object(
                 infeasible=True,
             ),
             _make_scatter_object(
-                non_best_trials,
+                info.non_best_trials_with_values,
                 hovertemplate="%{text}<extra>Feasible Trial</extra>",
                 dominated_trials=True,
             ),
             _make_scatter_object(
-                best_trials,
+                info.best_trials_with_values,
                 hovertemplate="%{text}<extra>Best Trial</extra>",
                 dominated_trials=False,
             ),
         ]
 
-    if n_dim == 2:
+    if info.n_targets == 2:
         layout = go.Layout(
             title="Pareto-front Plot",
-            xaxis_title=target_names[axis_order[0]],
-            yaxis_title=target_names[axis_order[1]],
+            xaxis_title=info.target_names[info.axis_order[0]],
+            yaxis_title=info.target_names[info.axis_order[1]],
         )
     else:
         layout = go.Layout(
             title="Pareto-front Plot",
             scene={
-                "xaxis_title": target_names[axis_order[0]],
-                "yaxis_title": target_names[axis_order[1]],
-                "zaxis_title": target_names[axis_order[2]],
+                "xaxis_title": info.target_names[info.axis_order[0]],
+                "yaxis_title": info.target_names[info.axis_order[1]],
+                "zaxis_title": info.target_names[info.axis_order[2]],
             },
         )
     return go.Figure(data=data, layout=layout)
 
 
+def _get_pareto_front_info(
+    study: Study,
+    target_names: Optional[List[str]] = None,
+    include_dominated_trials: bool = True,
+    axis_order: Optional[List[int]] = None,
+    constraints_func: Optional[Callable[[FrozenTrial], Sequence[float]]] = None,
+    targets: Optional[Callable[[FrozenTrial], Sequence[float]]] = None,
+) -> _ParetoFrontInfo:
+    if axis_order is not None:
+        warnings.warn(
+            "`axis_order` has been deprecated in v3.0.0. "
+            "This feature will be removed in v5.0.0. "
+            "See https://github.com/optuna/optuna/releases/tag/v3.0.0.",
+            DeprecationWarning,
+        )
+
+    if targets is not None and axis_order is not None:
+        raise ValueError(
+            "Using both `targets` and `axis_order` is not supported. "
+            "Use either `targets` or `axis_order`."
+        )
+
+    if constraints_func is not None:
+        feasible_trials = []
+        infeasible_trials = []
+        for trial in study.get_trials(states=(TrialState.COMPLETE,)):
+            if all(map(lambda x: x <= 0.0, constraints_func(trial))):
+                feasible_trials.append(trial)
+            else:
+                infeasible_trials.append(trial)
+        best_trials = _get_pareto_front_trials_by_trials(feasible_trials, study.directions)
+        if include_dominated_trials:
+            non_best_trials = _get_non_pareto_front_trials(feasible_trials, best_trials)
+        else:
+            non_best_trials = []
```

Comment:
```python
import optuna


def objective(trial):
    x = trial.suggest_float("x", 0, 5)
    y = trial.suggest_float("y", 0, 3)
    v0 = 4 * x ** 2 + 4 * y ** 2
    v1 = (x - 5) ** 2 + (y - 5) ** 2
    if x > 3:
        trial.set_user_attr("constraint", (1,))
    else:
        trial.set_user_attr("constraint", (0,))
    return v0, v1


def constraints(trial):
    return trial.user_attrs["constraint"]


study = optuna.create_study(
    sampler=optuna.samplers.NSGAIISampler(constraints_func=constraints),
    directions=["minimize", "minimize"],
)
study.optimize(objective, n_trials=100, n_jobs=1)

optuna.visualization.plot_pareto_front(
    study,
    constraints_func=constraints,
    include_dominated_trials=False,
)
```

I ran the optimization and visualized the pareto front plot to get the following figure:

<img width="987" alt="Screen Shot 2022-03-23 at 14 48 45" src="https://user-images.githubusercontent.com/5164000/159632394-632de63e-2718-4558-affc-442db5e8675a.png">

It looks like dominated trials are included in the figure.

I want ask you whether this is expected. I don't understand why they are plotted even `non_best_trials` is constructed from `feasible_trials`.

Reply:
Do you mean grey trials by "dominated trials"? In the current implementation, `infeasible_trials` are first excluded from all trials, then remaining ones are classified as `best` and `non_best`.
There will be some approaches for classifying trials:

- Exclude infeasible ones first, then classify feasible ones into best and non-best ones. (current approach)
- Exclude infeasible ones first, then classify feasible ones into best and non-best ones. Also, if `include_dominated_trials=False`, remove "dominated" infeasible trials.
- Exclude non-best ones first, then classify best ones by their feasibility. I think this is not a good idea, because feasible trials may be excluded just because they are dominated only by infeasible ones.

I understand that choice of the approach can be controversial.
Anyway, I think this PR itself should not change the behavior because this is for refactoring.

Reply:
Thank you for taking the time to explain, it totally makes sense. I assumed the third approach but I didn't care `feasible trials may be excluded just because they are dominated only by infeasible ones`.

I also agree with you that this PR should focus on the refactoring.